### ISSUE TITLE
[SEC] json-rpc - JINT is silently cast to uint if argument is declared as uint

### ISSUE THREAD

**Issue by tintinweb** [status-im/nimbus-eth2#1671]
//This is kind of confusing behavior and I don't really know why this is happening. Maybe someone can help out?

## Description

The json-rpc route handler parses json and unpacks it into the arguments declared for the method. The unpack routine appears to silently cast a JINT to `uint64` if the argument for the method handler is declared or resolves to `uint`. This is kind of unexpected and may lead to errors or make it harder to detect error conditions.

For example, consider this API method declaration which takes an argument `epoch: Epoch`. Epoch is declared as `uint`. Now, if we issue a request with a negative epoch for that rpc method, instead of returning an error, the method will proceed with the `uint(epoch)`.

https://github.com/status-im/nim-beacon-chain/blob/c76305f824fbc6a61eca5bbb9d21002eb6c515e3/beacon_chain/validator_api.nim#L370-L380

We are requesting a negative slot which we would expect the method to error. However, the method returns a result.

```
{
"jsonrpc":"2.0",
"method":"get_v1_validator_duties_proposer",
"params":[-9223372036854775807],
"id":2
}
  
```

response:
```
{"jsonrpc":"2.0","id":2,"result":[{"public_key":"b207f470bd025e6578b59a38238dad634509bb936cf142e9d15dbed8326b19e027783daa048b918baa6181d93b8e1c58","slot":32},{"public_key":"acd1541f7f8692c5ceb7f074b5eea44159b8b8b289fd9606babeed5ff83ec5c8256c408fdefc071ff10830b4055f0868","slot":33},{"public_key":"903f1b76b3615c2ae8d198b20f941f6e14a55016bbbee70237e14a79915d76252b8b392bb0463dc00b351ce325854889","slot":34},{"public_key":"affea88521d2afe737100358dde512c67f43e20f9d41e4d4c06ab39ce1b53568eaa6a9a0ec64a1800dbb70522ffa2082","slot":35},{"public_key":"97f0e58ca586a6b1d04eb470de9b8641b730f529a4806f6eade0355e1a777c95409a7ac735b697369957633ba3aecb80","slot":36},{"public_key":"ae28b9839f395b325da60a76234ad1d4050946954d619102fe9c8d96566d1fbc47f1dc64ce39cac0a5ac44153bb4c445","slot":37},{"public_key":"b6b67275658ae042cea502dfa9d30d7d36b5e0da33e70e9f49e4f692a8b921fc6930179c6ee354bf6940b5b610629498","slot":38},{"public_key":"a4fd5172645b8b3bba75d080dc915d6a0147a7f65e0e231c71e631167e2f0039ee1caeb74d219fcd31dc0d9f6cb7d229","slot":39},{"public_key":"a42b8a15bbeb83942dd9a9ed84b7bec374124005b233af4908e3f0c29d62537f4f24fa5fb7ffffd23ecfe751cae38220","slot":40},{"public_key":"a9ce9557824782bb3c97944002a112cab3d79299663cbd0964205a0cc86a8a9581130bc142fbecc2c5eef963c85cd2dd","slot":41},{"public_key":"964c841027f60cc95f174e18ab24ab7491549a6100de32d26341943c8c72e1df13e55b795b9e41e2d088f97fb663bd82","slot":42},{"public_key":"98ce2dde04e5be0d31b3c9c497c48c3853a65e457e03c38c442d2f63ef61a7885cfbd62fcaddca7369228cc9f6e8a747","slot":43},{"public_key":"a83c7679ccf803cbef1b96cb66d6b9f6ed8c46c67cc14655654e8a6aae34d81c90e83d381e79a219eaab0eb89c5892f2","slot":44},{"public_key":"adc89e67e9e084717f3c828a0e3bab50c8d21c6ea701853a3bf50c21cd02495a787619818addc8a004f4cbc57df2d5ed","slot":45},{"public_key":"81382c64c067bbbec9d0517fa0e3aff3bd1144f4d7382d5b8de7ccb7635a3ca8328db85d4759be74469a9a6360fe9aff","slot":46},{"public_key":"b6dad880afdecc2792f4081c814ff699745d6090a4476f4e7105c16a8864133a830a1e1bc90b600261f62bc18f0a8218","slot":47},{"public_key":"adc957c165e9dabb499f90c0a0f15a639ff7ff130ca081dc77a0e19dc7d5ba16c8b8f1c8dc8deb0e0151f6cc4a943099","slot":48},{"public_key":"875a842e8c9ad94ec9fe487775dc8c8aca6d1aa19ae1f0afdf6e3cda77dbcc6fc58e842de5707617e75076c34e282df7","slot":49},{"public_key":"8198729d3abec8e4d18a2ebb15dcc3d8527ba49c12fe9339e026f98f5fc269593d4cf09d593692f91d4ba37c90de1b30","slot":50},{"public_key":"84b1427894fd88bd98f42b99dcdd6e964a307fa9a0496cc93126e32929a954b0d3b6d84926ce63924aaeaeb21e81115c","slot":51},{"public_key":"b84ec93c2da55a04304c4bb24f7f85798ecd96ac2fc537391abb99c5e20e4b230ce5e741117087007da987f190ca3480","slot":52},{"public_key":"a4b8859facff2d8b48bf22e4e8e61ccf1fd70ca536ecbff3fca8d52fefd533e794d455d601c3a3fd59cf197d294e605c","slot":53},{"public_key":"ace846c616545230501faf89d443a1345bfb0cbad97467acdf4e5ec31a1b0d25ec7067ea2bd3c650cc3a7a97fee1edfe","slot":54},{"public_key":"acbf89d5dca0f6822e3aeec1a66cf57a389684250a163b2f86d82c437db914a073ec3f7e9b1547c6ccc826dcc31e4852","slot":55},{"public_key":"a4daea2eeeb9776582d640e574ed95fc16b325576cee055a7a9712cf84bb079113c9dc867456a719d9abbf28e7fbc140","slot":56},{"public_key":"90d72562a4b3310e30250004772b8f7ea8644d6c1c9cd26108e8766de3c23ef5c1e73b25133ef33b0bacbc376f82609e","slot":57},{"public_key":"b33dc34a3356f2f9a13ed4f81d6c230bad199b44575420b7d783ba59bf65bb2522bd3410e1699e70ab52d8fbbab82061","slot":58},{"public_key":"b944c194e718caab7fd33466507e68cd1b8e93e4f39c56656228f2c6edfeb833c64f9e3eb3bc3f23a83fb0da9373af30","slot":59},{"public_key":"abf493bc71074b21cdf1637b9815b4f19e433e29b42ccdc40d2a659c82e3a8fe64460afb57289b661bb23e9833e176d9","slot":60},{"public_key":"87084743d208145e8d30f30ac49ec443bf112d5b748c8aa69ebbbf00b45e74bd831561397d0f7cf7d391c96e7f8fc0b0","slot":61},{"public_key":"8cc8175cc4ade4fba9e1c2cd9c3569b8635278f1ee67dee81fcf373063893dc7ac6b93fa4666890f1729fd4ad07396e6","slot":62},{"public_key":"b7f0f98660518beaf68abad8fe6934acaa8b53fd1c1c20e38122c8e32d71dd5560fabfcdd57be7280881fd86ca5c5d3d","slot":63}],"error":null}

```

Checking the console output we can see a large positive value for epoch.

```
DBG 2020-09-16 16:30:14.205+02:00 get_v1_validator_duties_proposer           topics="valapi" tid=4074869 file=validator_api.nim:373 epoch=9223372036854775809
```

Now, what is weird is that we would be expecting an exception to be raised here: `newException(CatchableError, "Requesting way ahead of the current head") `. 
https://github.com/status-im/nim-beacon-chain/blob/c76305f824fbc6a61eca5bbb9d21002eb6c515e3/beacon_chain/validator_api.nim#L42-L49

But this is not what is happening and it is unclear to me why assuming that epoch is a large uint it should throw. Maybe someone from the dev team can help out here? 🙏 

## Exploit Scenario

not directly an exploit scenario but may aide in bypassing security checks and generally cover error conditions.

## Mitigation Recommendation

raise an exception when the source value (neg. int) does not fit into the target type (uint).
validate input before passing it along.

**Comment by mratsim** [comment]
This is incidentally fixed by #1810 while trying to fix #1663.

![image](https://user-images.githubusercontent.com/22738317/95207593-4f596d80-07e8-11eb-82a9-e49ce706a22a.png)

The conversion should still error out earlier than that check though.

**Comment by tintinweb** [comment]
(review 1994ffe5a0454c0395c4c216b8368434676defaa)

LGTM 

* Note: not behind insecure as per decision #1821
